keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,935 @@
+      to KEP (this should be a link to the KEP location in
+      kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+      [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+      links to mailing list discussions/SIG meetings, relevant
+      PRs/issues, release notes
+
+## Summary
+
+There are two types of volumes that are getting created after scheduling a pod onto a node:
+- [ephemeral inline
+  volumes](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
+- persistent volumes with [delayed
+  binding](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode)
+  (`WaitForFirstConsumer`)
+
+In both cases the Kubernetes scheduler currently picks a node without
+knowing whether the storage system has enough capacity left for
+creating a volume of the requested size. In the first case, `kubelet`
+will ask the CSI node service to stage the volume. Depending on the
+CSI driver, this involves creating the volume. In the second case, the
+`external-provisioner` will note that a `PVC` is now ready to be
+provisioned and ask the CSI controller service to create the volume
+such that is usable by the node (via
+[`CreateVolumeRequest.accessibility_requirements`](https://kubernetes-csi.github.io/docs/topology.html)).
+
+If these volume operations fail, pod creation gets stuck. The
+operations will get retried and might eventually succeed, for example
+because storage capacity gets freed up or extended. What does not
+happen is that the pod is re-scheduled to some other node which has
+enough storage capacity.
+
+A new API for exposing storage capacity currently available via CSI
+drivers and a scheduler enhancement that uses this information will
+reduce the risk of that happening.
+
+## Motivation
+
+### Goals
+
+* Define an API for exposing information about storage that is
+  flexible enough for a variety of use cases and that can be extended
+  later on.
+
+* Expose capacity information at the semantic
+  level that Kubernetes currently understands, i.e. in a way that
+  Kubernetes can compare capacity against the requested size of
+  volumes. This has to work for local storage, network-attached
+  storage and for drivers where the capacity depends on parameters in
+  the storage class.
+
+* Increase the chance of choosing a node for which volume creation
+  will succeed by tracking the currently available capacity available
+  through a CSI driver and using that information during pod
+  scheduling.
+
+### Non-Goals
```

Comment:
Changes to the scheduler also means impact on cluster autoscaler. Do we want to support autoscaling for storage plugins that implement this? We initially didn't for local volumes, but it's also becoming a popular ask now. cc @kubernetes/sig-autoscaling-proposals